File: problem-set/chapter2-2_river-crossing/brute-force.py
import numpy as np

dis_list = []
solution_list = []

def time_accmulation(time:np.ndarray, status:np.ndarray, torch:bool, dis=0, solution=[]):
    
    if status.all() == 1:
        dis_list.append(dis)
        solution_list.append(solution)
        return
    

    if torch == 0:
        for i in range(len(status)):
            if status[i] == 0:
                for j in range(i+1, len(status)):
                    if status[j] == 0:

                        status_copy = status.copy()
                        status_copy[i] = 1
                        status_copy[j] = 1

                        solution_copy = solution.copy()
                        solution_copy.append([time[i],time[j]])

                        time_accmulation(time, status_copy, 1, dis + max(time[i], time[j]), solution_copy)
                        
                
    if torch == 1:
        for i in range(len(status)):
            if status[i] == 1:
                
                status_copy = status.copy()
                status_copy[i] = 0

                solution_copy = solution.copy()
                solution_copy.append([time[i]])

                time_accmulation(time, status_copy, 0, dis + time[i], solution_copy)



if __name__ == '__main__':
    time = np.array([1,2,5,10,20])
    status = np.zeros(5, dtype=bool)
    torch = 0

    time_accmulation(time, status, torch)

    print(min(dis_list))
    print(solution_list[dis_list.index(min(dis_list))])

    

File: problem-set/chapter2-2_river-crossing/brute-force.cpp
#include <iostream>
#include <vector>
#include <algorithm>

std::vector<int> dis_list;

void time_accumulation(const std::vector<int>& time, std::vector<bool> status, bool torch, int dis = 0) {

    if (std::all_of(status.begin(), status.end(), [](bool s) { return s; })) {
        dis_list.push_back(dis);
        return;
    }

    if (torch == 0) {
        for (size_t i = 0; i < status.size(); ++i) {
            if (status[i] == 0) {
                for (size_t j = i + 1; j < status.size(); ++j) {
                    if (status[j] == 0) {

                        std::vector<bool> status_copy = status;
                        status_copy[i] = true;
                        status_copy[j] = true;

                        time_accumulation(time, status_copy, true, dis + std::max(time[i], time[j]));
                    }
                }
            }
        }
    }

    if (torch == 1) {
        for (size_t i = 0; i < status.size(); ++i) {
            if (status[i] == 1) {

                std::vector<bool> status_copy = status;
                status_copy[i] = false;

                time_accumulation(time, status_copy, false, dis + time[i]);
            }
        }
    }
}

int main() {
    std::vector<int> time = { 1, 2, 5, 10 };
    std::vector<bool> status(4, false);
    bool torch = false;

    time_accumulation(time, status, torch);

    for (int dis : dis_list) {
        std::cout << dis << " ";
    }
    std::cout << std::endl;

    return 0;
}
